bundles/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/handler/ICloudAccountBridgeHandler.java

```diff
@@ -86,21 +100,155 @@ public void handleCommand(ChannelUID channelUID, Command command) {
         }
     }
 
+    @SuppressWarnings("null")
     @Override
     public void initialize() {
         logger.debug("iCloud bridge handler initializing ...");
-        iCloudDeviceInformationCache = new ExpiringCache<>(CACHE_EXPIRY, () -> {
-            try {
-                return connection.requestDeviceStatusJSON();
-            } catch (IOException e) {
-                logger.warn("Unable to refresh device data", e);
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+
+        if (authState != AuthState.WAIT_FOR_CODE) {
+            authState = AuthState.INITIAL;
+        }
+
+        Callable<?> initialAuthentication = () -> callApiWithRetryAndExceptionHandling(() -> {
+            logger.debug("Dummy call for initial authentication.");
+            return null;
+        });
+        this.scheduler.schedule(initialAuthentication, 0, TimeUnit.SECONDS);
```

Comment:
If I understand it correctly, you need to sequentially ensure that initial authentication is done, then schedule refresh job. Perhaps it would be better to call initialAuthentication **from** the refresh job when not yet authenticated? This would also ensure that it's retried if the network call will fail during an authentication attempt. I didn't check if there is any other retry mechanism in place, but if not, this would improve reliability and at the same time solve the problem of multiple async jobs.

Reply:
Additionally, with current approach the future returned by `schedule()` should be stored, so it can be cancelled by `dispose()`.

Reply:
The flow here is a bit tricky. The refreshJob is only scheduled, if the authentication was successful:

```
        if (authState == AuthState.AUTHENTICATED) {
```

The problem is that the initial authentication call triggers the request for a 2-FA code. This request must not be automatically retried. Otherwise multiple codes could be requested. Since this is a oneshot job, I do not see see the need to cancel it.

Reply:
Yes you are right, this has to be done in the async task.

Reply:
The only question left is what will happen if the authentication will fail because of network issues, for example if internet connectivity is unavailable when the binding is started?

One consequence of the one-shot approach could be that authentication will not be retried at all, so that the binding will stay offline permanently. This requires users to manually stop and start the binding in such cases, so this is a reliability issue.

I haven't checked in detail if this is actually the case, just wanted to mention it. Even if we have this problem, I don't see it as a showstopper, this can be improved further later.

Reply:
I think this could happen. I do not like the current approach for 2-FA auth as part of init at all. It males the init complex and error prone. I‘m thinking of implementing a servlet or similar to provide a extra UI for it